MonoClass cleanup: array rank accessors, sizes union to MonoClass subtypes

lambdageek opened the pull request:
* Distribute the `sizes` union among the `MonoClass` subtypes:

  `element_size`, `generic_param_token` go to `MonoClassArray`,
  `MonoClassGenericParam`, respectively.

  `class_size` goes to each of `MonoClassDef`, `MonoClassGtd` and `MonoClassGenericInst`.

  This leaves `MonoClassPointer` without a sizes field.

* Move `rank` from `MonoClass` to `MonoClassArray`
  replace code like `if (klass->rank) ...` by `if (mono_class_is_array (klass))`
  replace use of `klass->rank` by `mono_class_get_array_rank`

lambdageek commented:
/cc @vargaz @kumpera

vargaz commented:
klass->rank is used in some perf critical code like allocators/sgen, so it might make sense to keep it in MonoClass.

lambdageek commented:
There are many places that were doing `if (klass->rank) { ... }` - those all got replaced by `if (mono_class_is_array (klass)) { ... }`  where `mono_class_is_array` is an inline function that expands to an `klass->kind == MONO_CLASS_ARRAY` which shoudl be just as fast as what was there before.

There are some places in `sgen-mono.c` where we have `mono_mb_emit_icon (mb, MONO_STRUCT_OFFSET (MonoClass, sizes));`   and a place in `method-to-ir.c` where we emit `MONO_EMIT_NEW_LOAD_MEMBASE_OP (cfg, OP_LOADU1_MEMBASE, rank_reg, klass_reg, MONO_STRUCT_OFFSET (MonoClass, rank));`  but in both of those cases we know statically that we're working with an array, so I changed them to `MONO_STRUCT_OFFSET (MonoClassArray, element_size)` and `MONO_STRUCT_OFFSET (MonoClassArray, rank)` respectively - so they're still just as fast as before.

Finally, there is some code in `sgen-mono.c` and `sgen-{new,tarjan}-bridge.c` that used to do `if (klass->rank == 1)` that I replaced `if (mono_class_is_array (klass) && mono_class_get_array_rank (klass) == 1)`.  This does seem like it might get slower.  We could run the GC benchmarks and compare.

kumpera commented:
I'm not sure on the sizes change as we don't save much space. Pointer types are really rare.

The distribution from the tests I'm using to measure footprint have the following:

MonoClassDef 26%
MonoClassGtd 1.34%
MonoClassGinst 63.95%
MonoClassGparam 2.59%
MonoClassArray 5.23%
MonoClassPointer 0.18%

As you can see, good changes take things aways fields from Def an Ginsts'.

kumpera commented:
In the case of the rank change. The are 2 things to consider.

We have a copy of that field in the vtable, we could change some JIT uses to that field instead.

Additionally, are we actually saving any size with it? rank is 1 byte and might end up just creating a hole in the struct.

I like having accessors for validation, but maybe the field movement is not helping us with the JIT.

lambdageek commented:
I changed the rank commit to just use accessors and left the field itself in MonoClass.

Moving rank to `MonoClassArray` might make sense in the future:
 1. if we can drop enough other small fields that we can save a whole word.
 2. if there's an easy way to access `MonoClass:class_kind` from IL (it's a 3 bit bitfield so it's much much easier to access `MonoClass:rank` which doesn't require knowing bit layout and emitting shifting instructions.)  (see method-to-ir.c `mini_emit_clastclass_inst()`)

I'm pretty ambivalent about this PR - if there's no space savings and it doesn't improve clarity or correctness, we can drop it.

vargaz commented:
I think moving rank doesn't really improve things too much, it would be better to concentrate on the int/pointer sized fields.

lambdageek commented:
There was not a lot of gain from this PR. Closing.
